### ISSUE TITLE
Bluetooth SN30

### ISSUE THREAD

**Issue by PD75** [libretro/Lakka-LibreELEC#595]
I followed the instruction to add an 8botdo SN30, but the Bluetooth stops working at startup and I keep having to recreate bluez.conf every reboot. I have dual boot with osmc and I have no problems what so ever with the controller there so no hardware issue, also tried RetroPie no problem there either.

The game works fine with keyboard din Lakka.

I have been unable to google any solution to this.

- Which version of Lakka are you using?
Lakka Verion: RPi2.arm-2.2.1
- What system hardware are you using?
RPi 3 B
- What did you do?
  - Instruction: http://www.lakka.tv/doc/8Bitdo-Wireless-Controller/
  - sudo Reboot
- What did you expect to happen?
  - Controller pair again after reboot
- What happened instead?
  - Not pairing
  - Redo from bluez.conf to pair

**Comment by fmorency** [comment]
Not sure if related but I'm also having some issue with this.

I have a pair of SN30 Pro controllers that I paired/trusted using the instruction in the LakkaTV wiki. Sometime the controllers pair again after reboot... Sometime they don't. When they don't, I can see in `bluetoothctl` that the controllers connect and disconnect right away. Not sure why...

To make it work again, I have to `remove` the devices using `bluetoothctl` and re-follow the instructions in the wiki. It works every time...

My hardware is an Odroid xu4q with the Odroid bluetooth adapter.

**Comment by Zewwy** [comment]
IIRC I had this issue as well...

Since I'm building another wicked lil system (a true portable SNES) and plan to use my SN30 Pro, I'll pair them up soon and report back.

**Comment by muis24** [comment]
I'm having issues with the SN30 Pro controllers as well since version 2.2 . Even after pairing, the input lags, only the up and down buttons work, or it doesn't work at all. The problems persist when using the wired USB mode. No issues when switching back to 2.1.1 . I can attach some logs, if necessary.

**Comment by tbocek** [comment]
I have the same issue as well and I updated even bluez from 5.43 to 5.50 in Lakka. The result was the same, sometimes the controller connects, sometimes I have to remove and connect manually. 

A workaround could be to do the reconnect automatically, similar to https://github.com/hooby3dfx/lakka-addons/blob/master/btautopair.sh

With RetroPie I had to set the connection mode to background (https://github.com/RetroPie/RetroPie-Setup/wiki/8bitdo-Controller), which I believe tries to reconnect in the background. Obviously the best solution would be to fix the connect/disconnect issue instead of such a workaround.

I have not tried Lakka 2.1.1 to see if bluez has the same issue, as this issue (https://github.com/libretro/Lakka-LibreELEC/issues/596) suggests that it might run with 2.1.1. However, if it does, then I'm even more puzzled what the issue is, as 2.1.1 has the same bluez version as 2.2

**Comment by slav0nic** [comment]
i have 2 8bitdo controllers and all works fine with latest nightly with https://github.com/libretro/Lakka-LibreELEC/commit/218d2bb1acb2a3915b7b1e7b0938c0e0b726ccfc fix or lakka9.2 branch without any extra scripts/services

**Comment by tbocek** [comment]
@slav0nic, thanks for the info, will try 9.2 in that case

**Comment by slav0nic** [comment]
@tbocek try latest master build http://nightly.builds.lakka.tv/2019-08-13/ (with bluez fix)

**Comment by fooness** [comment]
Any news or progress on this issue? I’m experiencing the same …

Pi 3B v1.2, Lakka-RPi2.RPi4.arm-2.3.1.img, 8bitdo firmware 4.20 …

**Comment by tbocek** [comment]
I also tryied 2.3.1, still have bluetooth connecting issues.

**Comment by fooness** [comment]
PS: When booting up Lakka, waiting for 1–2 minutes (also SSH service needs that time until it’s started), then powering on the gamepad, I get the following in my SSH session:

```
###########################################
# Lakka - The DIY retro emulation console #
# ...... visit http://www.lakka.tv ...... #
###########################################

Lakka (official): 2.3.1 (RPi2.RPi4.arm)
Lakka:~ # bluetoothctl
[NEW] Controller B8:27:EB:3D:A1:8F %h [default]
[NEW] Device E4:17:D8:46:65:6A 8Bitdo SFC30 GamePad
Agent registered
[8Bitdo SFC30 GamePad]#
```

So … the gamepad IS connected. But the retroarch interface shows no message about some registered controller and also doesn’t react to my key strokes.

**Comment by tbocek** [comment]
@slav0nic I can't try 9.2 on my XU4, as its not supported yet. I tried to update to bluez 5.51, but I have still the same issue. Settled now with a reconnect-script.

**Comment by slav0nic** [comment]
@tbocek strange, i tested my two 8bitdo gamepads with Lakka 2.3/PI4 -  works fine.
Are you used this gamepad only with lakka? (8bitdo can be paired only with 1 device).

Here is how i  pairing with pi4:
```
# removing old device for repairing (just for this test)
remove <addr>
scan on
# press start+x on SN30 with latest firmware, hold `select` 3sec
...
pair <addr>
trust <addr>
connect <addr>
```
checking info:
```
Lakka:~ # bluetoothctl 
...
[NEW] Device E4:17:D8:CF:9D:82 8BitDo SN30 gamepad
[NEW] Device E4:17:D8:26:8B:7C 8Bitdo SF30 Pro
Agent registered
[8BitDo SN30 gamepad]# info
Device E4:17:D8:CF:9D:82
	Name: 8BitDo SN30 gamepad
	Alias: 8BitDo SN30 gamepad
         ...
	Paired: yes
	Trusted: yes
	Connected: yes
```
checking  is pairing info stored correctly `ls -Rla  /storage/.cache/bluetooth/`, gamepad addr should be here.
Reboot Lakka, shutdown gamepad (hold start 3 sec), after Lakka started - enable gamepad by short  press `start`. Gamepad connected without any problems. 
AFAIK pairing info will be lost if you connect 8bitdo with other device (android phone for example)

**Comment by fooness** [comment]
@slav0nic How should `~/.storage/.cache/bluetooth/bluez.conf` look like? Mine is alwas empty, even after successfully pairing, trusting, and connecting to my SFC30 …

Is it possible to just add the two controllers MAC addresses there? Is it only possible for one MAC address? I already wondered, why the file is always empty.

**Comment by slav0nic** [comment]
@fooness, AFAIK fresh install don't have this config file (i'm not lakka dev, but it general bluez pkg with some patches and few changes in default config https://github.com/libretro/Lakka-LibreELEC/blob/1d82beebfd4b617acbcde0d4198c16f562cb00fe/packages/network/bluez/package.mk#L81-L85). 
Pairing information stored in `/storage/.cache/bluetooth/<bt_receiver_addr>/<paired_device_addr[1..n]>` (so after pairing 2 controlles you will see here 2 devices). I pairing 2 devices only via `bluetoothctl` command without editing any config files.

If you mean  `/storage/.cache/services/bluez.conf `, it's used only for enable bluetooth service and as `EnvironmentFile` https://github.com/libretro/Lakka-LibreELEC/blob/1d82beebfd4b617acbcde0d4198c16f562cb00fe/packages/network/bluez/system.d/bluetooth.service#L5 .

**Comment by franzbischoff** [comment]
Version 2.3.1 still cleaning the pairing list after reboot

**Comment by dmrzzz** [comment]
Same problem on Pi 4 with Lakka-RPi2.RPi4.arm-2.3.2.img.gz, using an older SNES30 (but upgraded to 8bitdo firmware v4.20).

I can trigger the problem without a full reboot, by restarting the bluetooth service.

1. Initial pairing as @slav0nic described, except I'm powering on with just Start (not Start+X). Controller works fine.

2. Hold Start to shut the controller off, guaranteeing that what's about to happen is not the controller's fault.

3. Check info before restart:

```
Lakka:~ # bluetoothctl
[NEW] Controller B8:27:EB:92:6E:F8 %h [default]
[NEW] Device E4:17:D8:6E:0B:75 8Bitdo SNES30 GamePad
Agent registered
[bluetooth]# info E4:17:D8:6E:0B:75
Device E4:17:D8:6E:0B:75
	Name: 8Bitdo SNES30 GamePad
	Alias: 8Bitdo SNES30 GamePad
	Class: 0x00002508
	Icon: input-gaming
	Paired: yes
	Trusted: yes
	Blocked: no
	Connected: no
	LegacyPairing: no
	UUID: Human Interface Device... (00001124-0000-1000-8000-00805f9b34fb)
	UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)
	Modalias: usb:v2DC8p2840d0100
[bluetooth]# exit
```

4. Restart bluetooth, and observe that **Paired: yes mysteriously changes to Paired: no** while everything else is remembered.

```
Lakka:~ # systemctl restart bluetooth
Lakka:~ # bluetoothctl
[NEW] Controller B8:27:EB:92:6E:F8 %h [default]
[NEW] Device E4:17:D8:6E:0B:75 8Bitdo SNES30 GamePad
Agent registered
[bluetooth]# info E4:17:D8:6E:0B:75
Device E4:17:D8:6E:0B:75
	Name: 8Bitdo SNES30 GamePad
	Alias: 8Bitdo SNES30 GamePad
	Class: 0x00002508
	Icon: input-gaming
	Paired: no
	Trusted: yes
	Blocked: no
	Connected: no
	LegacyPairing: no
	UUID: Human Interface Device... (00001124-0000-1000-8000-00805f9b34fb)
	UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)
	Modalias: usb:v2DC8p2840d0100
```

`/storage/.cache/bluetooth/B8:27:EB:92:6E:F8/E4:17:D8:6E:0B:75/info` still exists with the following contents:

```
[General]
Name=8Bitdo SNES30 GamePad
Class=0x002508
SupportedTechnologies=BR/EDR;
Trusted=true
Blocked=false
Services=00001124-0000-1000-8000-00805f9b34fb;00001200-0000-1000-8000-00805f9b34fb;

[DeviceID]
Source=2
Vendor=11720
Product=10304
Version=256

[LinkKey]
Key=1B7F83492EEBCC63451735E440B2AD38
Type=4
PINLength=0
```

Is there some other parameter that's not getting saved here but should?

**Comment by sarcasticadmin** [comment]
I tried my hand at a work around and I cannot consistently get my devices to pair. They definitely do not persistent after a reboot. I am running 2.3.2 (Generic.x86_64) but previous was on 2.1(Generic.x86_64) and had the same kinds of problems.

Here's my gist for anyone whose interested:
https://gist.github.com/sarcasticadmin/4558eee3093e68b46374b2ea0d7ffd3e

What makes it additional problematic is that lakka doesnt seem to have all of the typically tools for interacting with bluetooth like `simple-agent`
`bt-agent` or `rfkill`. Additionally, `bluetoothctl` doesnt have the ability to disable color output (hard to parse with control and color characters) and is not very scriptable.

I think Ill be using a long USB cord for the time being.

**Comment by franzbischoff** [comment]
Solved?

**Comment by Ntemis** [comment]
Come to discord and find the latest nightly. Try it and feel free to open a new issu
